File: lua/test/app_entity_test.lua
-- App entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("github_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("AppEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:App(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["app"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:App(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:App(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = app_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "app." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_APP_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local app_ref01_ent = client:App(nil)
    local app_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.app"), "app_ref01"))
    app_ref01_data["delivery_id"] = setup.idmap["delivery01"]
    app_ref01_data["installation_id"] = setup.idmap["installation01"]

    local app_ref01_data_result, err = app_ref01_ent:create(app_ref01_data, nil)
    assert.is_nil(err)
    app_ref01_data = helpers.to_map(type(app_ref01_data_result) == 'table' and app_ref01_data_result.data_get and app_ref01_data_result:data_get() or app_ref01_data_result)
    assert.is_not_nil(app_ref01_data)
    assert.is_not_nil(app_ref01_data["id"])

    -- LIST
    local app_ref01_match = {}

    local app_ref01_list_result, err = app_ref01_ent:list(app_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(app_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(app_ref01_list_result),
      { id = app_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local app_ref01_data_up0_up = {
      id = app_ref01_data["id"],
      ["installation_id"] = setup.idmap["installation_id"],
    }

    local app_ref01_markdef_up0_name = "access_tokens_url"
    local app_ref01_markdef_up0_value = "Mark01-app_ref01_" .. tostring(setup.now)
    app_ref01_data_up0_up[app_ref01_markdef_up0_name] = app_ref01_markdef_up0_value

    local app_ref01_resdata_up0_result, err = app_ref01_ent:update(app_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local app_ref01_resdata_up0 = helpers.to_map(type(app_ref01_resdata_up0_result) == 'table' and app_ref01_resdata_up0_result.data_get and app_ref01_resdata_up0_result:data_get() or app_ref01_resdata_up0_result)
    assert.is_not_nil(app_ref01_resdata_up0)
    assert.are.equal(app_ref01_resdata_up0["id"], app_ref01_data_up0_up["id"])
    assert.are.equal(app_ref01_resdata_up0[app_ref01_markdef_up0_name], app_ref01_markdef_up0_value)

    -- REMOVE
    local app_ref01_match_rm0 = {
      id = app_ref01_data["id"],
    }
    local _, err = app_ref01_ent:remove(app_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local app_ref01_match_rt0 = {}

    local app_ref01_list_rt0_result, err = app_ref01_ent:list(app_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(app_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(app_ref01_list_rt0_result),
      { id = app_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function app_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/app/AppTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read app test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "app01", "app02", "app03", "app_manifest01", "app_manifest02", "app_manifest03", "delivery01", "delivery02", "delivery03", "application01", "application02", "application03", "installation01", "installation02", "installation03", "repository01", "repository02", "repository03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("GITHUB_TEST_APP_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["GITHUB_TEST_APP_ENTID"] = idmap,
    ["GITHUB_TEST_LIVE"] = "FALSE",
    ["GITHUB_TEST_EXPLAIN"] = "FALSE",
    ["GITHUB_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["GITHUB_TEST_APP_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end
  if idmap_resolved["installation_id"] == nil then
    idmap_resolved["installation_id"] = idmap_resolved["installation01"]
  end

  if env["GITHUB_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["GITHUB_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["GITHUB_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["GITHUB_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end

File: go/test/app_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestAppEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.App(nil)
		if ent == nil {
			t.Fatal("expected non-nil AppEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"app": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.App(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.App(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := appBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "app." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_APP_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		appRef01Ent := client.App(nil)
		appRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "app"}), "app_ref01"))
		appRef01Data["delivery_id"] = setup.idmap["delivery01"]
		appRef01Data["installation_id"] = setup.idmap["installation01"]

		appRef01DataResult, err := appRef01Ent.Create(appRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		appRef01Data = core.ToMapAny(entityData(appRef01DataResult))
		if appRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if appRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		appRef01Match := map[string]any{}

		appRef01ListResult, err := appRef01Ent.List(appRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		appRef01List, appRef01ListOk := appRef01ListResult.([]any)
		if !appRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", appRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(appRef01List), map[string]any{"id": appRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		appRef01DataUp0Up := map[string]any{
			"id": appRef01Data["id"],
			"installation_id": setup.idmap["installation_id"],
		}

		appRef01MarkdefUp0Name := "access_tokens_url"
		appRef01MarkdefUp0Value := fmt.Sprintf("Mark01-app_ref01_%d", setup.now)
		appRef01DataUp0Up[appRef01MarkdefUp0Name] = appRef01MarkdefUp0Value

		appRef01ResdataUp0Result, err := appRef01Ent.Update(appRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		appRef01ResdataUp0 := core.ToMapAny(entityData(appRef01ResdataUp0Result))
		if appRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if appRef01ResdataUp0["id"] != appRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if appRef01ResdataUp0[appRef01MarkdefUp0Name] != appRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", appRef01MarkdefUp0Name, appRef01ResdataUp0[appRef01MarkdefUp0Name])
		}

		// REMOVE
		appRef01MatchRm0 := map[string]any{
			"id": appRef01Data["id"],
		}
		_, err = appRef01Ent.Remove(appRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		appRef01MatchRt0 := map[string]any{}

		appRef01ListRt0Result, err := appRef01Ent.List(appRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		appRef01ListRt0, appRef01ListRt0Ok := appRef01ListRt0Result.([]any)
		if !appRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", appRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(appRef01ListRt0), map[string]any{"id": appRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func appBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "app", "AppTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read app test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse app test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"app01", "app02", "app03", "app_manifest01", "app_manifest02", "app_manifest03", "delivery01", "delivery02", "delivery03", "application01", "application02", "application03", "installation01", "installation02", "installation03", "repository01", "repository02", "repository03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_APP_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_APP_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_APP_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add installation_id alias for update test.
	if idmapResolved["installation_id"] == nil {
		idmapResolved["installation_id"] = idmapResolved["installation01"]
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
